scipy/stats/_discrete_distns.py

```diff
@@ -1262,6 +1263,133 @@ def _stats(self, alpha):
+    where
+    :math:`x_l = \max(0, n - m_2)`,
+    :math:`x_u = \min(n, m_1)`,
+    :math:`m_2 = N - m_1`,
+
+    .. math::
+
+        P_0 = \sum_{y=x_l}^{x_u} \binom{m_1}{y}\binom{m_2}{n-y}\omega^y,
+
+    and the binomial coefficients are defined as
+
+    .. math:: \binom{n}{k} \equiv \frac{n!}{k! (n - k)!}.
+
+    `fnch` uses the BiasedUrn package by Agner Fog with permission
+    for it to be distributed under SciPy's license.
+
+    The symbols used to denote the shape parameters (`M`, `n`, and `N`) are not
+    universally accepted; they are chosen for consistency with `hypergeom`.
+    The choice of corresponding mathematical symbols allows for
+    comparison with the references.
+
+    Note that Fisher's noncentral hypergeometric distribution is distinct
+    from Wallenius' noncentral hypergeometric distribution, which models
+    drawing a pre-determined `N` objects from a bin one by one.
+    When the odds ratio is unity, however, both distributions reduce to the
+    ordinary hypergeometric distribution.
+
+    %(after_notes)s
+
+    References
+    ----------
+    .. [1] Agner Fog, "Biased Urn Theory".
+           https://cran.r-project.org/web/packages/BiasedUrn/vignettes/UrnTheory.pdf
+
+    .. [2] "Fisher's noncentral hypergeometric distribution", Wikipedia,
+           https://en.wikipedia.org/wiki/Fisher's_noncentral_hypergeometric_distribution
+
+    %(example)s
+
+    """
+
+    def _get_support(self, M, n, N, odds):
+        N, m1, n = M, n, N  # follow Wikipedia notation
+        m2 = N - m1
+        x_min = np.maximum(0, n - m2)
+        x_max = np.minimum(n, m1)
+        return x_min, x_max
+
+    def _argcheck(self, M, n, N, odds):
+        M, n = np.asarray(M), np.asarray(n),
+        N, odds = np.asarray(N), np.asarray(odds)
+        cond1 = (M.astype(int) == M) & (M >= 0)
+        cond2 = (n.astype(int) == n) & (n >= 0)
+        cond3 = (N.astype(int) == N) & (N >= 0)
+        cond4 = odds > 0
+        cond5 = N <= M
+        cond6 = n <= M
+        return cond1 & cond2 & cond3 & cond4 & cond5 & cond6
+
+    def _rvs(self, M, n, N, odds, size=None, random_state=None):
+
+        def _rvs1(M, n, N, odds, length, random_state=None):
+            urn = _PyStochasticLib3()
+            rvs = urn.rvs_fisher(N, n, M, odds, length, random_state)
+            return rvs
+
+        # If all the shapes are scalar, we can generate all the random numbers
+        # in one call to rvs_fisher and reshape to the desired result.
+        # Otherwise, we need to resort to np.vectorize.
+        if (np.size(M) == 1 and np.size(n) == 1
+                and np.size(N) == 1 and np.size(odds) == 1):
+            rvs = _rvs1(M, n, N, odds,
+                        np.prod(size), random_state).reshape(size)
+        else:
+            fun = np.vectorize(_rvs1, excluded=['size', 'random_state'])
+            rvs = fun(M, n, N, odds, 1, random_state)
+        return rvs
+
+    def _pmf(self, x, M, n, N, odds):
+
+        @np.vectorize
```

Comment:
@mckib2 can we vectorize this and `_rvs`, too?

Reply:
How do you mean vectorize?  As in make a ufunc or manually loop in Cython?  We could do either

Reply:
Yes, either : )
Whichever is easier, I imagine, unless the other is way faster.